### ISSUE TITLE
Smiley parsing can break inline attachments

### ISSUE THREAD

**Issue by Oldiesmann** [SimpleMachines/SMF#3179]
While testing my fixes for the inline attachment issues, I discovered a new problem...

I have an image attachment with an ID of 8.

The bbcode for displaying this inline results in the following HTML:

``` html
<a href=".../index.php?action=dlattach;topic=782.0;attach=8;image" id="link_8" onclick="return expandThumb(8);"><img src=".../index.php?action=dlattach;topic=782.0;attach=9;image" alt="" id="thumb_8"></a>
```

The problem is, `8)` is also a smiley code. When parsing smileys, SMF sees that and plops the img code for that smiley right in the middle of the onclick attribute, resulting in this mangled code:

``` html
<a href=".../index.php?action=dlattach;topic=782.0;attach=8;image" id="link_8" onclick="return expandThumb(<img src=".../Smileys/fugue/cool.gif" alt="8&#41;" title="Cool" class="smiley">;"><img src=".../index.php?action=dlattach;topic=782.0;attach=9;image" alt="" id="thumb_8"></a>
```

We need to make sure we're not parsing smiley code inside an HTML attribute somewhere. This is an extremely minor bug and I'm not sure how likely it is to occur, but it's something we should try to fix before final.

**Comment by margarett** [comment]
I'm not sure if it's such a minor issue really. It should affect every thumb whose ID ends in 8, not just the 8 itself. Isn't it?

**Comment by Oldiesmann** [comment]
Possibly. I don't know for sure though.

**Comment by Kindred-999** [comment]
oof....    this is not actually a minor bug.   it's minor in EFFECT, but it will affect things as margarett suggests (I have seen the 8) trigger in the middle of a word and sentence before, as well as some of the other smileys - it's a PITA in text, but a potential for LOTS of complaints with attachments.

**Comment by margarett** [comment]
In parsesmileys (Subs.php) there is this:

> ```
>   // This smiley regex makes sure it doesn't parse smileys within code tags (so [url=mailto:David@bla.com] doesn't parse the :D smiley)
> ```

Plus some code which ends in a regex which should (I think :P) prevents smileys from being parsed in undesired places.
I know _niente_ of regex, but could we just make it ignore everything between
<a ...... ">
:question: 

edit: eh, looks like I got it, thanks to https://regex101.com/ and a lot of trial and error ;D
Let me test some more.

**Comment by margarett** [comment]
OK so, final findings after some time testing smileys back-and-forth.

a) it only happens with thumb ID = 8. I used str_replace to simulate several IDs, with or without "8"s and the only one which breaks is 8. I think I understand why but I don't want to risk saying :tongue: 
b) it can be fixed by changing

```
(?=[^[:alpha:]0-9]|$)
```

With

```
(?=[^[:alpha:]0-9]<a|$)
```

This will prevent smileys from being parsed inside "anchor" tags but _NOT_ inside the URL's description, so one can still do this:

```
[url=http://www.google.pt]LOL great 8)[/url]
```

And still see the smiley correctly parsed.

I would suggest @Oldiesmann, @MissAllSunday (or any one else who actually knows regex :tongue: ) to verify this before a PR is created (by myself or anyone else)

**Comment by Oldiesmann** [comment]
I'm not a regex person so I can't help... The best option would be to prevent parsing smileys of any kind inside the <a tag like that. In the URL text itself it doesn't matter too much - just as long as it doesn't break stuff like it currently does.

**Comment by margarett** [comment]
As far as I could test that's exactly what that change does ;) It prevents any smiley to be parsed inside <a.......
In other places they are parsed normally

**Comment by margarett** [comment]
Bump ;)

Should I create a PR for this or should we wait for further comments?

**Comment by MissAllSunday** [comment]
Yeah that regex should work.

**Comment by FearThe1337** [comment]
Hmm this commit broke all smileys on my forum.
I took the time to dump the base64 encoded value of the regex both how it was before this commit and after. The regex after does not detect any smileys as shown in the script below:

``` php
$old_regex = base64_decode("fig/PD1bPjpcP1wuXHNceHtBMH1bXF0oKSpcXDtdfF4pKFw6bGlnaHRzdG9uZWR1c3RcOnxcOmRpYW1vbmRwaWNrYXhlXDp8XDpkaWFtb25kc2hvdmVsXDp8XDpkaWFtb25kYmxvY2tcOnxcOnllbGxvd2Zsb3dlclw6fFw6amFja29sYW50ZXJuXDp8XDpjaHJpc3RtYXNwaWdcOnxcOmRpYW1vbmRzd29yZFw6fFw6Z3JlZW5yZWNvcmRcOnxcOndhdGVyYnVja2V0XDp8XDpnb2xkZW5hcHBsZVw6fFw6Zm9ybWFsc2hlZXBcOnxcOm1vYnNwYXduZXJcOnxcOmRpYW1vbmRob2VcOnxcOmRpYW1vbmRheGVcOnxcOmdyYXNzeWRpcnRcOnxcOmNvb2tlZG1lYXRcOnxcOmxpZ2h0c3RvbmVcOnxcOmZpc2hpbmdyb2RcOnxcOnRvd2VsaWVkb2hcOnxcOmxhdmFidWNrZXRcOnxcOnN1Z2FyY2FuZXNcOnxcOm1pbGtidWNrZXRcOnxcOndvb2RlbmRvb3JcOnxcOmlyb25ibG9ja1w6fFw6Ym9va3NoZWxmXDp8XDpyZWRmbG93ZXJcOnxcOmdvbGRibG9ja1w6fFw6d29ya2JlbmNoXDp8XDp0cm9sbGZhY2VcOnxcOnNsaW1lYmFsbFw6fFw6c25vd2JhbGxcOnxcOm1pbmVjYXJ0XDp8XDptdXNocm9vbVw6fFw6cGFpbnRpbmdcOnxcOmZ1cm5hY2VcOnxcOmxlYXRoZXJcOnxcOmNyZWVwZXJcOnxcOmRpYW1vbmRcOnxcOnNhZGRsZVw6fFw6em9tYmllXDp8XDpidWNrZXRcOnxcOmdyYXZlbFw6fFw6YXBwbGVcOnxcOnN0aWNrXDp8XDpzdWdhclw6fFw6bGF1Z2hcOnxcOnN0ZXZlXDp8XDpoZWFydFw6fFw6YW5nZWxcOnxcOnBhcGVyXDp8XDpjaGVzdFw6fFw6c2hlZXBcOnxcOmRpcnRcOnxcOnNpZ25cOnxcOmZpcmVcOnxcOmlyb25cOnxcOmdvbGRcOnxcOmZpc2hcOnxcOmNha2VcOnxcOmJvYXRcOnxcOmJvbmVcOnxcOmJvb2tcOnxcOmVnZ1w6fFw6Y293XDp8XDpib3dcOnxcOnBpZ1w6fFw6c2Z0XDp8XDp0bnRcOnxcPlw6XCh8Jmd0O1w6XCh8XD9cP1w/fFw+XDpEfCZndDtcOkR8XDonXCh8XDomIzM5O1wofFw6XC1cKnxcOlwtXFx8XDpcLVh8XDpcLVxbfFw6XDpcKXxcOlwpfFw6RHw7RHxPMHxcOlB8OFwpfFw6b3xcOlwofDtcKSkoPz1bXls6YWxwaGE6XTAtOV18JCl+dQ==");
$new_regex = base64_decode("fig/PD1bPjpcP1wuXHNceHtBMH1bXF0oKSpcXDtdfF4pKFw6bGlnaHRzdG9uZWR1c3RcOnxcOmRpYW1vbmRwaWNrYXhlXDp8XDpkaWFtb25kc2hvdmVsXDp8XDpkaWFtb25kYmxvY2tcOnxcOnllbGxvd2Zsb3dlclw6fFw6amFja29sYW50ZXJuXDp8XDpjaHJpc3RtYXNwaWdcOnxcOmRpYW1vbmRzd29yZFw6fFw6Z3JlZW5yZWNvcmRcOnxcOndhdGVyYnVja2V0XDp8XDpnb2xkZW5hcHBsZVw6fFw6Zm9ybWFsc2hlZXBcOnxcOm1vYnNwYXduZXJcOnxcOmRpYW1vbmRob2VcOnxcOmRpYW1vbmRheGVcOnxcOmdyYXNzeWRpcnRcOnxcOmNvb2tlZG1lYXRcOnxcOmxpZ2h0c3RvbmVcOnxcOmZpc2hpbmdyb2RcOnxcOnRvd2VsaWVkb2hcOnxcOmxhdmFidWNrZXRcOnxcOnN1Z2FyY2FuZXNcOnxcOm1pbGtidWNrZXRcOnxcOndvb2RlbmRvb3JcOnxcOmlyb25ibG9ja1w6fFw6Ym9va3NoZWxmXDp8XDpyZWRmbG93ZXJcOnxcOmdvbGRibG9ja1w6fFw6d29ya2JlbmNoXDp8XDp0cm9sbGZhY2VcOnxcOnNsaW1lYmFsbFw6fFw6c25vd2JhbGxcOnxcOm1pbmVjYXJ0XDp8XDptdXNocm9vbVw6fFw6cGFpbnRpbmdcOnxcOmZ1cm5hY2VcOnxcOmxlYXRoZXJcOnxcOmNyZWVwZXJcOnxcOmRpYW1vbmRcOnxcOnNhZGRsZVw6fFw6em9tYmllXDp8XDpidWNrZXRcOnxcOmdyYXZlbFw6fFw6YXBwbGVcOnxcOnN0aWNrXDp8XDpzdWdhclw6fFw6bGF1Z2hcOnxcOnN0ZXZlXDp8XDpoZWFydFw6fFw6YW5nZWxcOnxcOnBhcGVyXDp8XDpjaGVzdFw6fFw6c2hlZXBcOnxcOmRpcnRcOnxcOnNpZ25cOnxcOmZpcmVcOnxcOmlyb25cOnxcOmdvbGRcOnxcOmZpc2hcOnxcOmNha2VcOnxcOmJvYXRcOnxcOmJvbmVcOnxcOmJvb2tcOnxcOmVnZ1w6fFw6Y293XDp8XDpib3dcOnxcOnBpZ1w6fFw6c2Z0XDp8XDp0bnRcOnxcPlw6XCh8Jmd0O1w6XCh8XD9cP1w/fFw+XDpEfCZndDtcOkR8XDonXCh8XDomIzM5O1wofFw6XC1cKnxcOlwtXFx8XDpcLVh8XDpcLVxbfFw6XDpcKXxcOlwpfFw6RHw7RHxPMHxcOlB8OFwpfFw6b3xcOlwofDtcKSkoPz1bXls6YWxwaGE6XTAtOV08YXwkKX51");
$smileyPregReplacements = "dummy";


$message = " :)  :laugh:  ;)  :D  ;D  >:(  :(  :o  8) ";
$message = preg_replace_callback($old_regex,
    function ($matches) use ($smileyPregReplacements)
    {
        echo "Match[old]: " . $matches[1] . "\n";
        return $matches[1];
    }, $message);


$message = " :)  :laugh:  ;)  :D  ;D  >:(  :(  :o  8) ";
$message = preg_replace_callback($new_regex,
    function ($matches) use ($smileyPregReplacements)
    {
        echo "Match[new]: " . $matches[1] . "\n";
        return $matches[1];
    }, $message);
```

Output:

```
Match[old]: :)
Match[old]: :laugh:
Match[old]: ;)
Match[old]: :D
Match[old]: ;D
Match[old]: >:(
Match[old]: :(
Match[old]: :o
Match[old]: 8)
```

tldr; pre-commit worked fine, with this commit it ignores all smileys.

**Comment by margarett** [comment]
:open_mouth: 
I did test that as we were discussing the issue. Will double check asap.

**Comment by margarett** [comment]
Indeed this is true :disappointed: 
I don't know what happened between my tests and the PR but the fact is that it totally messes smileys. I'm dang sure I tested smileys back and forth but... Anyway I already deleted those files so there is nothing to do about that anymore :(
@Oldiesmann please revert the PR and reopen this issue.

edit: Very strange. Some smileys do work but only if don't have a space in front of them. Others don't work at all (eg: :laugh:)

edit2: eh, I don't have a "laugh" smiley :tongue: No wonder it didn't work :bulb:

**Comment by FearThe1337** [comment]
@margarett just took some time to look at the regex expression.
I modified the regex as PHP used, except filled in some custom smileys.

Given the original regex:

``` PCRE
(?<=[>:\?\.\s\x{A0}[\]()*\\;]|^)(8D|:smile:|:\))(?=[^[:alpha:]0-9]|$)
```

Based on my understanding it does the following:

### Group 1:

```
(?<=[>:\?\.\s\x{A0}[\]()*\\;]|^)
```

Ensures that the character before the next part (The smiley match) either:
- does not exist (start of string)
- -or- is one of the given characters: 
  -      > : ? . [ ]() \* \ ;
  -      is equal to \s, which is a whitespace character. (Space, tab, carriage return, newline, vertical tab, form feed)
  -     Equal to the character 0xA0 (Non-breaking space)

### Group 2:

```
(8D|:smile:|:\))
```

Matches the string literals that identify a smiley, so, e.g.

```
8D, :smile: and :)
```

Not much to say about this part.

### Group 3:

```
(?=[^[:alpha:]0-9]|$)
```

This group ensures that the character after the smiley part is either
- a character that is not numeric and not alphabetic  (So, not: a-z a-Z 0-9)
- or, there is no character after the smiley. (end of string)

### Modification

The way/location where you added <a makes the last block different when there is a character after the smiley.
It will only match in this case when there is a character that fits condition 1 of the last block (not: a-z a-Z 0-9) followed by <a (literally).
This modification makes it so that:

``` php
    $does_not_match_1 = "8D "; // Character after the smiley, but no <a
    $does_not_match_2 = "8D <"; // No 'a' so <a does not match
    $does_not_match_2 = "8D <b"; // 'b' is not an a.
    $does_not_match_2 = "8D1<a"; // '1' is not allowed after the smiley part '8D'
    $does_not_match_2 = "8Da<a"; // 'b' is not allowed after the smiley part '8D'
    $does_match_1     = "8D"; // End of string matched
    $does_match_1     = "8D <a"; // [non-alpha(numeric]<a matched
    $does_match_1     = "8D*<a";// [non-alpha(numeric]<a matched
```

#### Personal conclusion:

I'm not 100% sure if it is ideal to fix this in the regular expression regardless.

#### Additional minor note:

It seems the regular expression is applied per line in the post after parsing the bb-codes.
so a smiley and then directly a new-line also counts as the end of the string/input.

**Comment by Oldiesmann** [comment]
Reverted the fix since it broke things... Reopening the issue for further analysis/discussion.
